### ISSUE TITLE
Can we use the features extracted by the R(2+1)D-34 caffe model as input for a classifier built in Keras?

### ISSUE THREAD

**Issue by ThatIndianCoder** [facebookresearch/VMZ#86]
For one of my projects, I would like to build a simple feed forward neural network in Keras. I was wondering if it is possible to use the features extracted by the given script written in caffe2 as input for a keras model.

**Comment by bjuncek** [comment]
I'm sure it is, you'd simply need to write the dataloader that takes those features in for Keras - unfortunately, I can't really help with that :(


---
I've done something similar for pytorch, and it worked

**Comment by ThatIndianCoder** [comment]
Thank you for the tip. I will look into it.

**Comment by rsomani95** [comment]
> I've done something similar for pytorch, and it worked

@bjuncek do you intend on sharing the code you wrote for doing this in PyTorch? I think a lot of people would find that very helpful.

**Comment by bjuncek** [comment]
@daniel-j-h wrote a nifty little script here

>>>Hey folks we just did that for the r(2+1)d 34-layer model here: https://github.com/moabitcoin/ig65m-pytorch


I'll make my repo public as well tonight, as it might have a bit more context for other models :)

Best,
Bruno

**Comment by rsomani95** [comment]
@bjuncek this repo might be exactly what I'm looking for! Thank you for the reference :)

To summarise:
1. VMZ defined and trained these models in Caffe2, exported and shared them as `.pkl` files, but to use them via PyTorch, you need to define the same model in PyTorch and copy over the weights layer by layer from the `.pkl` file -- and that's what https://github.com/moabitcoin/ig65m-pytorch does. 
2. In order to train on a custom dataset, we could create a torch `Dataset` using https://github.com/irhum/R2Plus1D-PyTorch/blob/master/dataset.py and fine-tune the loaded model. 

Is my understanding correct?

**Comment by daniel-j-h** [comment]
The `.pkl` files are the (key, value) blobs for the trained Caffe2 model weights.

In the [convert.py](https://github.com/moabitcoin/ig65m-pytorch/blob/791b64797252cf0d36c1fbbe505d73d1df0c1f64/convert.py) tool in [ig65m-pytorch](https://github.com/moabitcoin/ig65m-pytorch) we manually create the matching architecture in PyTorch (check convert source for slight model differences and how we fix them), and then copy over the named blobs into the appropriate PyTorch modules. We double check for tensor sizes, dtype matches, and that we have a 1:1 mapping between blobs and PyTorch model - with enough care to details this works!

We also provide an [extract.py](https://github.com/moabitcoin/ig65m-pytorch/blob/791b64797252cf0d36c1fbbe505d73d1df0c1f64/extract.py) script showing how to use model and weights on your custom dataset and e.g. on your webcam. We'll work on more convenience tools in the coming weeks.

**Comment by rsomani95** [comment]
@daniel-j-h thanks for the clarification. Excellent repo!

**Comment by bjuncek** [comment]
@rsomani95 yeah, that's pretty much the idea. I'd also take a look into the torchvision datasets, as I've tested it with those and can't really guarantee performance for other dataloaders (unfortunately video training can be quite data-dependent). In addition, I found them to be the simplest to set up as well.

ref: https://github.com/pytorch/vision/blob/master/torchvision/datasets/kinetics.py 



and thanks :)
